src/relax/backend/vm/rewrite_cuda_graph.cc

```diff
@@ -0,0 +1,433 @@
+    };
+    while (!task_stack.empty()) {
+      const VarNode* node = task_stack.back();
+      task_stack.pop_back();
+      f_sink(node);
+    }
+    // remove nodes from the graph
+    for (const auto* node : nodes_to_remove) {
+      graph->bindings.erase(node);
+    }
+    auto f_is_removed = [&](const VarNode* node) { return nodes_to_remove.count(node); };
+    graph->nodes.erase(std::remove_if(graph->nodes.begin(), graph->nodes.end(), f_is_removed),
+                       graph->nodes.end());
+    graph->alloc_tensors.erase(
+        std::remove_if(graph->alloc_tensors.begin(), graph->alloc_tensors.end(), f_is_removed),
+        graph->alloc_tensors.end());
+  }
+
+  Function EmitStaticAllocations(BlockBuilderNode* builder, StaticRegion* region) {
+    builder->BeginBindingBlock();
+    Array<Expr> alloc_tensors;
+    Array<StructInfo> alloc_tensors_struct_info;
+    for (const auto& var : scope_.graph.alloc_tensors) {
+      region->alloc_tensor_to_index.emplace_back(var, alloc_tensors.size());
+      alloc_tensors.push_back(builder->Emit(scope_.graph.bindings.at(var)->value));
+      alloc_tensors_struct_info.push_back(
+          Downcast<StructInfo>(GetRef<Var>(var)->struct_info_.value()));
+    }
+    auto output = builder->Emit(Tuple(alloc_tensors));
+    auto block = builder->EndBlock();
+    auto func_body = builder->Normalize(SeqExpr({block}, output));
+    auto func = Function({}, func_body, TupleStructInfo(alloc_tensors_struct_info));
+    return func;
+  }
+
+  Function EmitStaticGraph(BlockBuilderNode* builder, StaticRegion* region) {
+    Var allocs{"allocs", region->alloc_func->ret_struct_info};
+    builder->BeginBindingBlock();
+    for (int i = 0; i < static_cast<int>(scope_.graph.alloc_tensors.size()); ++i) {
+      const VarNode* tensor = scope_.graph.alloc_tensors[i];
+      const Expr alloc = builder->Normalize(TupleGetItem(allocs, i));
+      builder->EmitNormalized(VarBinding(GetRef<Var>(tensor), alloc));
+    }
+    for (const VarNode* node : scope_.graph.nodes) {
+      const VarBindingNode* binding = scope_.graph.bindings.at(node);
+      builder->EmitNormalized(GetRef<VarBinding>(binding));
+    }
+    auto void_output = builder->Emit(Tuple(Array<Expr>{}));
+    auto block = builder->EndBlock();
+    auto func_body = builder->Normalize(SeqExpr({block}, void_output));
+    auto func = Function({allocs}, func_body, TupleStructInfo(Array<StructInfo>{}));
+    return func;
+  }
+
+  void SummarizeStaticRegion() {
+    PruneOutputNodes(&scope_.graph, outputs_);
+
+    // Find all bindings other than alloc_tensors to emit
+    std::vector<VarBinding> bindings_to_emit;
+    bool has_kernel_launch = false;
+    for (const auto* var : scope_.graph.nodes) {
+      auto it = scope_.graph.bindings.find(var);
+      const VarBindingNode* binding = it->second;
+      if (const CallNode* call = binding->value.as<CallNode>();
+          call != nullptr && call->op.as<GlobalVarNode>()) {
+        has_kernel_launch = true;
```

Comment:
Do we want to check if `call->op` corresponds to a PrimFunc in the IRModule? I’m thinking about the case of Relax cross-function calls. Looks like if `call->op` means a Relax function, the static region will also be created.